udfs/cloudbuild.yaml

```diff
@@ -14,19 +14,143 @@
 #
 
 steps:
-###########################################################
-# Step 1: Run UDF Unit Tests
-###########################################################
-- name: python:3.8.0-slim
-  entrypoint: 'bash'
-  args: ['udfs/tests/run.sh', '--pip_install_before_run']
-  env: ['SHORT_SHA=$SHORT_SHA']
-###########################################################
-# Step 2: Deploy if all tests pass
-###########################################################
+  ###########################################################
+  # Step 0: Install npm packages
+  ###########################################################
+- name: 'node'
+  id: 'step_0_install_npm_packages'
+  entrypoint: 'npm'
+  args:
+    - install
+  dir: js_libs
+  ###########################################################
+  # Step 1: Build all js libraries for BigQuery UDFs
+  ###########################################################
+- name: 'node'
+  id: 'step_1_build_bq_js_libs'
+  entrypoint: 'npm'
+  args:
+    - run-script
+    - build-all-libs
+  dir: js_libs
+  waitFor:
+    - 'step_0_install_npm_packages'
+  ###########################################################
+  # Step 2: Copy libs to GCS bucket
+  ###########################################################
 - name: gcr.io/google.com/cloudsdktool/cloud-sdk
-  entrypoint: 'bash'
-  args: ['release/build.sh', '$BRANCH_NAME', '$_PR_NUMBER']
+  id: 'step_2_copy_js_to_gcs'
+  entrypoint: 'gsutil'
+  args:
+    - cp
+    - builds/*
+    - ${_JS_BUCKET}
+  dir: js_libs
+  waitFor:
+    - 'step_1_build_bq_js_libs'
+  ###########################################################
+  # Step 3: Create python virtual environment
+  ###########################################################
+- name: python:3.8.0-slim
```

Comment:
All these cloud build steps are extremely inefficient (slow / expensive) and not easy for a contributor to replicate locally (for example if they do not have python 3.8 installed locally).

Instead define a dockerfile and build / cache it (for speed). This way the contributor can just build your dockerfile and use it to run tests. This will make it very easy to contribute from a browser in github codespaces once it goes GA (or from any machine w/ docker installed).

You can read about the advantages of this approach in my other PR [here](https://github.com/GoogleCloudPlatform/bigquery-utils/pull/182/files#diff-98292c6c7bafaade39268394353ece80d647970e0f65601653882704735a29aeR202).

To give you and idea of what it would look like here (of course I'd also recommend adding more python linters to catch common issues):

Dockerfile.ci
```Dockerfile
FROM python:3.8.0-slim
WORKDIR /ci
COPY . /ci/
RUN pip3 install -r tests/requirements.txt
# This can be overriden in cloud build
ENTRYPOINT ["tests/udf_test_utils.py"] 
```

cloudbuild.yaml
```yaml
# steps 1 & 2 ...
- name: 'hadolint/hadolint'  # optional
  entrypoint: '/bin/hadolint'
  args:
    - 'Dockerfile.ci'
  id: 'lint-ci-docker-image'
- name: 'gcr.io/kaniko-project/executor:latest'
  waitFor:
    - 'lint-ci-docker-image'
  args:
    - '--dockerfile=Dockerfile.ci'
    - '--destination=gcr.io/$PROJECT_ID/bq_udf_ci'
    - '--cache=true'
  id: 'build-ci-image-from-kaniko-cache'
- name: gcr.io/$PROJECT_ID/bq_udf_ci
  id: 'create_test_datasets'  # use Dockerfile's ENTRYPOINT
  args:
    - --create_test_datasets
  env:
    - SHORT_SHA=${SHORT_SHA}
  waitFor:
    - 'build-ci-image-from-kaniko-cache'
  ###########################################################
  # Step 6: Create UDF signatures with empty bodies to prevent
  # UDF creation errors in next step when one UDF depends
  # on another UDF.
  ###########################################################
- name: gcr.io/$PROJECT_ID/bq_udf_ci
  id: 'create_udf_signatures'
  entrypoint: 'pytest'   # Overide entrypoint locally
  args:
    - --workers
    - '100'
    - tests/create_udf_signatures.py
  env:
    - SHORT_SHA=${SHORT_SHA}
  waitFor:
    - 'copy_js_to_gcs'
    - 'create_test_datasets'
 # etc...
```